lib/Dialect/Sim/SimOps.cpp

```diff
@@ -397,6 +397,223 @@ LogicalResult PrintFormattedProcOp::canonicalize(PrintFormattedProcOp op,
+                                SmallVectorImpl<OpFoldResult> &results) {
+  if (auto constCond = dyn_cast_or_null<IntegerAttr>(adaptor.getCondition())) {
+    if (constCond.getValue().isAllOnes()) {
+      // Strip constant true condition.
+      getConditionMutable().clear();
+      return success();
+    }
+    // Never enabled, fold to tie-offs.
+    if (getNumResults() > 0) {
+      results.append(adaptor.getTieoffsAttr().begin(),
+                     adaptor.getTieoffsAttr().end());
+      return success();
+    }
+  }
+  return failure();
+}
+
+LogicalResult TriggeredOp::canonicalize(TriggeredOp op,
+                                        PatternRewriter &rewriter) {
+  if (op.getNumResults() > 0)
+    return failure();
+
+  bool isDeadOrEmpty = false;
+
+  auto *bodyBlock = &op.getBodyRegion().front();
+  isDeadOrEmpty = bodyBlock->without_terminator().empty();
+
+  if (!isDeadOrEmpty && !!op.getCondition())
+    if (auto cstCond = op.getCondition().getDefiningOp<hw::ConstantOp>())
+      isDeadOrEmpty = cstCond.getValue().isZero();
+
+  if (!isDeadOrEmpty)
+    return failure();
+
+  rewriter.eraseOp(op);
+  return success();
+}
+
+// --- TriggerSequenceOp ---
+
+LogicalResult TriggerSequenceOp::inferReturnTypes(
+    MLIRContext *context, std::optional<Location> location, ValueRange operands,
+    DictionaryAttr attributes, OpaqueProperties properties, RegionRange regions,
+    SmallVectorImpl<Type> &inferredReturnTypes) {
+  // Create N results matching the type of the parent trigger, where N is the
+  // specified length of the sequence.
+  auto lengthAttr =
+      properties.as<TriggerSequenceOp::Properties *>()->getLength();
+  uint32_t len = lengthAttr.getValue().getZExtValue();
+  Type trigType = operands.front().getType();
+  inferredReturnTypes.resize_for_overwrite(len);
+  for (size_t i = 0; i < len; ++i)
+    inferredReturnTypes[i] = trigType;
+  return success();
+}
+
+LogicalResult TriggerSequenceOp::verify() {
+  if (getLength() != getNumResults())
+    return emitOpError("specified length does not match number of results.");
+  return success();
+}
+
+LogicalResult TriggerSequenceOp::fold(FoldAdaptor adaptor,
+                                      SmallVectorImpl<OpFoldResult> &results) {
+  // Fold trivial sequences to the parent trigger.
+  if (getLength() == 1 && getResult(0) != getParent()) {
+    results.push_back(getParent());
+    return success();
+  }
+  return failure();
+}
+
+LogicalResult TriggerSequenceOp::canonicalize(TriggerSequenceOp op,
+                                              PatternRewriter &rewriter) {
+  if (op.getNumResults() == 0) {
+    rewriter.eraseOp(op);
+    return success();
+  }
+
+  // If the current op can be inlined into the parent,
+  // leave it to the parent's canonicalization.
+  if (auto parentSeq = op.getParent().getDefiningOp<TriggerSequenceOp>()) {
+    if (parentSeq == op) {
+      op.emitWarning("Recursive trigger sequence.");
```

Comment:
Isn't this a failure?  Shouldn't it be covered by the verifier?  What about multi-op sequences which are circular?

Reply:
In practice a cyclic trigger graph is almost certainly a bug. But theoretically it is still well-formed: It has no root event, thus it is never invoked. There is no good way of handling this right now, but that should change with the soon to be added `sim.never` op, which just creates a constant "dead" trigger.

Multi-op circular graphs will be opportunistically handled by the flattening DFS. I'm not sure it is worth spending any effort on making sure _all_ cycles are detected and replaced.